File: C++/tests/49_test.cpp
#include <gtest/gtest.h>
#include <solution_tmpl.h>
#include <string>
#include <vector>
using namespace std;

class GroupAnagrams : public ::testing::Test {
protected:
  Solution solution;
};

void check_equal(vector<vector<string>> result, vector<vector<string>> want) {
  for (size_t i = 0; i < result.size(); i++) {
    for (size_t j = 0; j < result[i].size(); j++) {
      string target = result[i][j];
      bool found = false;
      for (size_t r = 0; r < want[i].size(); r++) {
        if (want[i][r] == target) {
          found = true;
          break;
        }
      }
      if (!found) {
        ADD_FAILURE_AT(__FILE__, __LINE__)
            << "Mismatch at result[" << i << "][" << j << "]: " << target
            << " not found in want[" << i << "]";
      }
    }
  }
}

TEST_F(GroupAnagrams, SameLengths) {
  vector<string> strs = {"eat", "tea", "tan", "ate", "nat", "bat"};
  vector<vector<string>> want = {
      {"bat"}, {"nat", "tan"}, {"ate", "eat", "tea"}};
  vector<vector<string>> result = solution.groupAnagrams(strs);

  check_equal(result, want);
}

TEST_F(GroupAnagrams, EmptyStrings) {
  vector<string> strs = {""};
  vector<vector<string>> want = {{""}};
  vector<vector<string>> result = solution.groupAnagrams(strs);

  check_equal(result, want);
}

File: JavaScript/tests/49.test.js
import { describe, test } from "node:test";
import { groupAnagrams } from "../src/49.js";

/**
 * @param {string[][]} result
 * @param {string[][]} expected
 */
function assertEqualAnagrams(result, expected) {
	if (result.length !== expected.length) {
		throw new Error(
			`Assertion failed!\n Expect: ${JSON.stringify(expected)}\n Actual: ${JSON.stringify(result)}`,
		);
	}

	for (let i = 0; i < expected.length; i++) {
		let found = false;

		for (let j = 0; j < result.length; j++) {
			if (expected[i].length !== result[j].length) {
				continue;
			}

			found = expected[i].every((str) => result[j].includes(str));
			if (found) {
				break;
			}
		}

		if (!found) {
			throw new Error(
				`Assertion failed!\n Expect: ${JSON.stringify(expected)}\n Actual: ${JSON.stringify(result)}`,
			);
		}
	}
}

const cases = [
	{
		strs: ["eat", "tea", "tan", "ate", "nat", "bat"],
		want: [["bat"], ["nat", "tan"], ["ate", "eat", "tea"]],
	},
	{
		strs: ["a"],
		want: [["a"]],
	},
	{
		strs: [""],
		want: [[""]],
	},
];

describe("49. Group Anagrams", () => {
	for (let i = 0; i < cases.length; i++) {
		test(`case ${i + 1}`, () => {
			const result = groupAnagrams(cases[i].strs);
			assertEqualAnagrams(result, cases[i].want);
		});
	}
});
